File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> visited;
        int res = 0;
        int n = s.size();
        int i = 0, j = 0;

        while(j < n) {
            if(visited.contains(s[j])) {
                i++;
                j = i;
                visited.clear();
            } else {
                res = max(res, (j-i+1));
                visited.insert(s[j]);
                j++;
            }
        }

        return res;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        i, j = 0, 0
        res = 0
        visited = set()
        
        while j < len(s):
            while s[j] in visited and i < j:
                visited.remove(s[i])
                i += 1
            
            res = max(res, j - i + 1)
            visited.add(s[j])
            j += 1
        
        return res